ambry-replication/src/main/java/com/github/ambry/replication/PartitionLeaderInfo.java

```diff
@@ -0,0 +1,155 @@
+    } finally {
+      rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Remove a partition from the map of leader partitions. This method is thread safe.
+   * @param partitionName name of the partition to be removed
+   */
+  public void removePartition(String partitionName) {
+    // Read-write lock avoids contention from threads adding new leaders (addPartition()) and threads updating existing leader partitions (refreshPeerLeadersForAllPartitions())
+    rwLock.writeLock().lock();
+    try {
+      peerLeaderReplicasByPartition.remove(partitionName);
+    } finally {
+      rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Refreshes the list of remote leaders for all leader partitions by querying the latest information from RoutingTableSnapshots of all data centers.
+   * This method is thread safe.
+   */
+  public void refreshPeerLeadersForAllPartitions() {
+
+    // Read-write lock usage: Avoids contention between threads doing the following activities:
+    // 1. Adding new leaders (in addPeerLeadersByPartition())
+    // 2. Removing old leaders (in removePartition())
+    // 3. Refreshing remote leader set for existing leaders (current method).
+    // Explanation for point 3: Multiple threads from different cluster change handlers (we have one cluster change handler for each DC) can trigger onRoutingTableUpdate() in parallel which calls this method to refresh leader partitions.
+    // We need to make sure that the sequence of gathering remote leaders (from RoutingTableSnapshot of each DC) and updating the map is an atomic operation.
+
+    rwLock.writeLock().lock();
+    try {
+      for (Map.Entry<String, Set<ReplicaId>> entry : peerLeaderReplicasByPartition.entrySet()) {
+        String partitionName = entry.getKey();
+        ReplicaId localLeaderReplica = storeManager.getReplica(partitionName);
+        PartitionId partition = localLeaderReplica.getPartitionId();
+        Set<ReplicaId> previousRemoteLeaderReplicas = entry.getValue();
+        Set<ReplicaId> currentRemoteLeaderReplicas =
+            new HashSet<>(partition.getReplicaIdsByState(ReplicaState.LEADER, null));
+        currentRemoteLeaderReplicas.remove(localLeaderReplica);
+        if (!previousRemoteLeaderReplicas.equals(currentRemoteLeaderReplicas)) {
+          peerLeaderReplicasByPartition.put(partitionName, currentRemoteLeaderReplicas);
+        }
+      }
+    } finally {
+      rwLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Checks if a remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+   * @param partitionName name of local leader partition
+   * @param replicaId remote replica to be checked
+   * @return true if remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+   */
+  public boolean isPeerReplicaLeaderForPartition(String partitionName, ReplicaId replicaId) {
```

Comment:
How will this method be used?

Reply:
We want to use this method in inter-colo ReplicaThreads to filter leader and non-leader (standby, offline, bootstrap) replicas from their lists of remote replicas.